src/Microsoft.OData.Core/UriParser/ODataPathInfo.cs

```diff
@@ -44,6 +46,19 @@ public ODataPathInfo(ODataPath odataPath)
                         this.targetEdmType = collectionType.ElementType.Definition;
                     }
                 }
+
+                // If the last segment is an OperationSegment for a bound operation and the targetNavigationSource is null,
+                // We use the targetNavigationSource for the previous segment.
+                // e.g People/My.Function.GetPeopleWithDogs()
+                if (this.targetNavigationSource == null && lastSegment is OperationSegment operationSegment)

```

Comment:
This is where you'd check if `previous` is null to avoid doing a lot of work for nothing...